python/ql/src/semmle/python/frameworks/Tornado.qll

```diff
@@ -0,0 +1,543 @@
+
+        /** Gets a reference to the `full_url` method. */
+        private DataFlow::Node full_url(DataFlow::TypeTracker t) {
+          t.startInAttr("full_url") and
+          result = instance()
+          or
+          exists(DataFlow::TypeTracker t2 | result = full_url(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `full_url` method. */
+        DataFlow::Node full_url() { result = full_url(DataFlow::TypeTracker::end()) }
+
+        private class AdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
+          override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+            // Method access
+            nodeTo.(DataFlow::AttrRead).getObject() = nodeFrom and
+            nodeFrom = instance() and
+            nodeTo in [full_url()]
+            or
+            // Method call
+            nodeTo.asCfgNode().(CallNode).getFunction() = nodeFrom.asCfgNode() and
+            nodeFrom in [full_url()]
+            or
+            // Attributes
+            nodeFrom = instance() and
+            exists(DataFlow::AttrRead read | nodeTo = read and read.getObject() = nodeFrom |
+              read.getAttributeName() in [
+                  // str / bytes
+                  "uri", "path", "query", "remote_ip", "body",
+                  // Dict[str, List[bytes]]
+                  "arguments", "query_arguments", "body_arguments",
+                  // dict-like, https://www.tornadoweb.org/en/stable/httputil.html#tornado.httputil.HTTPHeaders
+                  "headers",
+                  // Dict[str, http.cookies.Morsel]
+                  "cookies"
+                ]
+            )
+          }
+        }
+      }
+    }
+  }
+
+  // ---------------------------------------------------------------------------
+  // routing
+  // ---------------------------------------------------------------------------
+  /** A sequence that defines a number of route rules */
+  SequenceNode routeSetupRuleList() {
+    exists(CallNode call | call = any(tornado::web::Application::ClassInstantiation c).asCfgNode() |
+      result in [call.getArg(0), call.getArgByName("handlers")]
+    )
+    or
+    exists(CallNode call |
+      call.getFunction() = tornado::web::Application::add_handlers().asCfgNode()
+    |
+      result in [call.getArg(1), call.getArgByName("host_handlers")]
+    )
+    or
+    result = routeSetupRuleList().getElement(_).(TupleNode).getElement(1)
+  }
+
+  /** A tornado route setup. */
+  abstract class TornadoRouteSetup extends HTTP::Server::RouteSetup::Range { }
+
+  /**
+   * A regex that is used to set up a route.
+   *
+   * Needs this subclass to be considered a RegexString.
+   */
+  private class TornadoRouteRegex extends RegexString {
+    TornadoRouteSetup setup;
+
+    TornadoRouteRegex() {
+      this instanceof StrConst and
+      DataFlow::localFlow(DataFlow::exprNode(this), setup.getUrlPatternArg())
+    }
```

Comment:
We might be able to ask for a LocalSourceNode here.

Reply:
Yep, just rewrote it to match DjangoRouteRegex